relays/substrate/src/headers_initialize.rs

```diff
@@ -88,50 +83,153 @@ async fn do_initialize<SourceChain: Chain, TargetChain: Chain>(
 	Ok(initialization_tx_hash)
 }
 
-/// Prepare initialization data for the headers-bridge pallet.
+/// Prepare initialization data for the finality-verifier pallet.
 async fn prepare_initialization_data<SourceChain: Chain>(
 	source_client: Client<SourceChain>,
-	raw_initial_header: Option<Bytes>,
-	raw_initial_authorities_set: Option<Bytes>,
-	initial_authorities_set_id: Option<GrandpaAuthoritiesSetId>,
 ) -> Result<InitializationData<SourceChain::Header>, String> {
-	let source_genesis_hash = *source_client.genesis_hash();
-
-	let initial_header = match raw_initial_header {
-		Some(raw_initial_header) => SourceChain::Header::decode(&mut &raw_initial_header.0[..])
-			.map_err(|err| format!("Failed to decode {} initial header: {:?}", SourceChain::NAME, err))?,
-		None => source_client
-			.header_by_hash(source_genesis_hash)
-			.await
-			.map_err(|err| format!("Failed to retrive {} genesis header: {:?}", SourceChain::NAME, err))?,
-	};
-
-	let raw_initial_authorities_set = match raw_initial_authorities_set {
-		Some(raw_initial_authorities_set) => raw_initial_authorities_set.0,
-		None => source_client
-			.grandpa_authorities_set(source_genesis_hash)
-			.await
-			.map_err(|err| {
-				format!(
-					"Failed to retrive {} authorities set at genesis header: {:?}",
-					SourceChain::NAME,
-					err
-				)
-			})?,
-	};
-	let initial_authorities_set =
-		GrandpaAuthoritiesSet::decode(&mut &raw_initial_authorities_set[..]).map_err(|err| {
-			format!(
-				"Failed to decode {} initial authorities set: {:?}",
+	// In ideal world we just need to get best finalized header and then to read GRANDPA authorities
+	// set (`pallet_grandpa::CurrentSetId` + `GrandpaApi::grandpa_authorities()`) at this header.
+	//
+	// But now there are problems with this approach - `CurrentSetId` may return invalid value. So here
+	// we're waiting for the next justification, read the authorities set and then try to figure out
+	// the set id with bruteforce.
+	let mut justifications = source_client
+		.subscribe_justifications()
+		.await
+		.map_err(|err| format!("Failed to subscribe to {} justifications: {:?}", SourceChain::NAME, err))?;
+
+	// Read next justification - the header that it finalizes will be used as initial header.
+	let justification = justifications.next().await.ok_or_else(|| {
+		format!(
+			"Failed to read {} justification from the stream: stream has ended unexpectedly",
+			SourceChain::NAME,
+		)
+	})?;
```

Comment:
Would it make sense to have a retry mechanism here?

Reply:
It's not about what this PR is about - if we'll ever change initialization approach from "just submit transaction" to "check that pallet is not initialized, submit transaction and wait until it's mined", then retry will make sense. This particular error can't happen in practice, because we need to miss 4096 justifications and we're reading just after subscription.